cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerServiceTest.java

```diff
@@ -0,0 +1,184 @@
+
+    // Call the sidecar again and make sure the same path was returned
+    File sameUnpackedDir = client.getUnpackedArtifactLocation(artifactId.toEntityId());
+    Assert.assertEquals(unpackedDir, sameUnpackedDir);
+
+    // Delete and recreate the artifact to update the last modified date
+    artifactRepository.deleteArtifact(artifactId);
+
+    Thread.sleep(1000);
+    Files.copy(appJarFile, newAppJarFile);
+    artifactRepository.addArtifact(artifactId, newAppJarFile);
+
+    File newUnpackDir = client.getUnpackedArtifactLocation(artifactId.toEntityId());
+
+    //Make sure the two paths arent the same and that the old one is gone
+    Assert.assertNotEquals(unpackedDir, newUnpackDir);
+    validateUnpackDir(newUnpackDir);
+  }
+
+  private void validateUnpackDir(File unpackedFile) {
+    // Make sure the directory exists
+    Assert.assertTrue(unpackedFile.exists());
+    Assert.assertTrue(unpackedFile.isDirectory());
+
+    // Make sure theres multiple files in the directory and one of them is a manifest
+    String[] fileNames = unpackedFile.list();
+    Assert.assertTrue(fileNames.length > 1);
+    Assert.assertTrue(Arrays.stream(fileNames).anyMatch(s -> s.equals("META-INF")));
+  }
+
+  @Test
+  public void testArtifact() throws Exception {
+
+    LocationFactory locationFactory = getInjector().getInstance(LocationFactory.class);
+    ArtifactRepository artifactRepository = getInjector().getInstance(ArtifactRepository.class);
+    ArtifactLocalizerClient client = getInjector().getInstance(ArtifactLocalizerClient.class);
+
+    Id.Artifact artifactId = Id.Artifact.from(Id.Namespace.DEFAULT, "some-task", "1.0.0-SNAPSHOT");
+    Location appJar = AppJarHelper.createDeploymentJar(locationFactory, TaskWorkerServiceTest.TestRunnableClass.class);
+    File appJarFile = new File(tmpFolder.newFolder(),
+                               String.format("%s-%s.jar", artifactId.getName(), artifactId.getVersion().getVersion()));
+    File newAppJarFile = new File(tmpFolder.newFolder(),
+                                  String.format("%s-%s-copy.jar", artifactId.getName(),
+                                                artifactId.getVersion().getVersion()));
+    Files.copy(Locations.newInputSupplier(appJar), appJarFile);
+    appJar.delete();
+    artifactRepository.addArtifact(artifactId, appJarFile);
+
+    File artifactPath = client.getArtifactLocation(artifactId.toEntityId());
+
+    // Make sure the artifact was actually cached
+    Assert.assertTrue(artifactPath.exists());
+
+    // Call the sidecar again and make sure the same path was returned
+    File sameArtifactPath = client.getArtifactLocation(artifactId.toEntityId());
+    Assert.assertEquals(artifactPath, sameArtifactPath);
+
+    // Delete and recreate the artifact to update the last modified date
+    artifactRepository.deleteArtifact(artifactId);
+
+    // This sleep is needed to delay the file copy so that the lastModified time on the file is different
+    Thread.sleep(1000);
```

Comment:
Better not to sleep in unit-test

Reply:
Whoops, I didnt realize I can set the lastModified time explicitly. I've removed the sleep and used that instead.

Reply:
This sleep is needed to make sure the `lastModified` timestamp is different when this artifact is re-added to the artifact repo. If I remove this sleep the test becomes flaky because it depends on how long the files take to copy and how long the requests take to return. 

I can directly set the lastModified time using `file.setLastModified()` but that gets overwritten by `artifactRepository.addArtifact()` since it does a copy as part of its work. Is there a better way I can explicitly set the lastModified time in the artifact repo?

Reply:
You can change the file timestamp directly in the test instead of sleeping

Reply:
I tried that but that only modifies the timestamp for the file **before** it gets added to the artifact repository.

Once it gets added to the artifact store the file is copied so the new file ([see code](https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactStore.java#L693)) has a different lastModified time. Adding this sleep delays the `artifactRepository.addArtifact()` call which makes sure that the lastModified time will be different